yjit/src/codegen.rs

```diff
@@ -7061,15 +7062,38 @@ fn iseq_get_return_value(iseq: IseqPtr, captured_opnd: Option<Opnd>, ci_flags: u
     }
     match first_insn {
         YARVINSN_getlocal_WC_0  => {
-            // Only accept simple positional only cases for both the caller and the callee.
+            // Accept only cases where only positional arguments are used by both the callee and the caller.
+            // Keyword arguments may be specified by the callee or the caller but not used.
             // Reject block ISEQs to avoid autosplat and other block parameter complications.
-            if captured_opnd.is_none() && unsafe { rb_simple_iseq_p(iseq) } && ci_flags & VM_CALL_ARGS_SIMPLE != 0 {
-                let ep_offset = unsafe { *rb_iseq_pc_at_idx(iseq, 1) }.as_u32();
-                let local_idx = ep_offset_to_local_idx(iseq, ep_offset);
-                Some(IseqReturn::LocalVariable(local_idx))
-            } else {
-                None
+            if captured_opnd.is_some()
+                // Reject if block ISEQ is present
+                || block.is_some()
+                // Equivalent to `VM_CALL_ARGS_SIMPLE - VM_CALL_KWARG - has_block_iseq`
+                || ci_flags & (
+                      VM_CALL_ARGS_SPLAT
+                    | VM_CALL_KW_SPLAT
+                    | VM_CALL_ARGS_BLOCKARG
+                    | VM_CALL_FORWARDING
+                ) != 0
+                 {
+                return None;
+            }
+
+            let ep_offset = unsafe { *rb_iseq_pc_at_idx(iseq, 1) }.as_u32();
+            let local_idx = ep_offset_to_local_idx(iseq, ep_offset);
+
+            if unsafe { rb_simple_iseq_p(iseq) } {
+                return Some(IseqReturn::LocalVariable(local_idx));
+            } else if unsafe {  rb_iseq_only_kwparam_p(iseq) } {
+                // Inline if only positional parameters are used
+                if let Ok(i) = i32::try_from(local_idx) {
```

Comment:
There is an implicit check going on here that should be made explicit in comment. For callees that have expressions in the optional keywords like: `def foo(kw: itself) = nil`, we're relying on the instruction count check to filter them out. Note that depending on the nature of the expression the iseq you get for the callee is different. Compare to `def foo(kw: 3) = nil`, for example.

Reply:
I have to leave now but I will add this later this afternoon.

Reply:
I added a comment [here](https://github.com/Shopify/ruby/blob/9df72f8ec0137a0539506d06e1503563783fd9d5/yjit/src/codegen.rs#L7050-L7051). Let me know if it's clear enough.